ai/core/src/unitycatalog/ai/core/databricks.py

```diff
@@ -181,34 +150,45 @@ def wrapper(self, *args, **kwargs):
         for attempt in range(1, max_attempts + 1):
             try:
                 result = func(self, *args, **kwargs)
-                # for non-session related error in the result, we should directly return the result
-                if (
-                    isinstance(result, FunctionExecutionResult)
-                    and result.error
-                    and SESSION_EXCEPTION_MESSAGE in result.error
-                ):
-                    raise Exception(result.error)
+                if isinstance(result, FunctionExecutionResult) and result.error:
+                    if any(
+                        msg in result.error
+                        for msg in (SESSION_EXPIRED_MESSAGE, SESSION_CHANGED_MESSAGE)
+                    ):
+                        raise Exception(result.error)
+                    return result
                 return result
             except Exception as e:
                 error_message = str(e)
-                if SESSION_EXCEPTION_MESSAGE in error_message:
+                if any(
+                    msg in error_message
+                    for msg in (SESSION_EXPIRED_MESSAGE, SESSION_CHANGED_MESSAGE)
+                ):
                     if not self._is_default_client:
-                        refresh_message = f"Failed to execute {func.__name__} due to session expiration. Unable to automatically refresh session when using a custom client."
+                        refresh_message = (
+                            f"Failed to execute {func.__name__} due to session expiration. "
+                            "Unable to automatically refresh session when using a custom client."
+                            "Recreate the DatabricksFunctionClient with a new client to recreate "
+                            "the custom client session."
+                        )
```

Comment:
Actually probably it's possible to recreate a WorkspaceClient based on current client configs, but this can be a follow-up :)

Reply:
Yeah I think I'm going to do some testing on that one just to be sure... perhaps early next week